### ISSUE TITLE
ee('Filesystem')->exists($path) falsely reports files as not being present

### ISSUE THREAD

**Issue by jcogs-design** [ExpressionEngine/ExpressionEngine#3866]
**Description of the problem**
Encountered an issue where running a highly nested instance of JCOGS Image (i.e. applying up to 7 different transformations to an image through multiple instances of Image running sequentially). The processing was failing mid-way through. Upon digging through the logs, the reason was down to a point being reached where Image was not being able to verify the existance of the interim output images generated by earlier instances in the sequence. If Image cannot locate the input file nominated it (unsurprisingly) stops processing - which is why the chained transformation sequence failed. However on closer inspection it transpired the images that were being reported as not existing were in fact present. The issue was arising from *how* the presence of the files was being determined, not actually whether they were present or not.

Image uses (or probably better to say used) the ee Filesystem library call `ee('Filesystem')->exists($path)` to validate that a local source image was present. It did this to ensure maximum compatibilty with EE's new Flysystem based file system. It would appear that this library call is not capable of handingly rapidly updating files. Stepping through the sequence it appears the library call in EE7 actually uses a Flysystem library call to determine the existance of a local file - and so it might be that this issue is simply a reflection of deeper issues within Flysystem.

Replacing the ee call with the simpler php `is_readable($path)` call solved the issue (and marginally improved Image processing performance) - but it is unclear how well this php call would interact with EE setups that use more exotic Flysystem drivers etc.

**How To Reproduce**
A reliable method is to run the same highly nested Image call using version 1.3.10 of the add-on ... the template code for this is numbingly complicated to pick through and relies upon three source images to work, but is included here simply for reference. It might be that some other less opaque method can be found to test. The primary issue here is to resolve whether using this call is required when exotic Flysystem connectors are in use rather than the basic php / unix primitives - if so then this test needs to be improved to work with this kind of context. If not, then there is nothing that needs doing.

**Screenshots / Videos / Template Code**
<pre><code>{exp:jcogs_img:image:book src='{web_card_images:image_1}' height='450' allow_scale_larger='yes'}
    {exp:jcogs_img:image:black src='/media/site_images/black_square_200.webp' width='950' height='630' fit='distort' allow_scale_larger='yes' filename='black'}
        {exp:jcogs_img:image:shadow src='/media/site_images/white_square_200.webp' width='{book:width}' height='{book:height}' fit='distort' allow_scale_larger='yes' filename='shadow'}
            {exp:jcogs_img:image:frame src='{black:made}' width='200%' allow_scale_larger='yes' filename='frame'}
                {exp:jcogs_img:image:add_shadow src='{frame:made}' watermark='{shadow:made}|0,0|100|center,center|15,15' filename='add_shadow' allow_scale_larger='yes'}
                    {exp:jcogs_img:image:add_image src='{add_shadow:made}' watermark='{book:made}|0,0|100|center,center|-15,-15' filename='add_image' allow_scale_larger='yes'}
                        {exp:jcogs_img:image:combined src='{black:made}' watermark='{add_image:made}|0,0|100|center,center' filename='{book:name_orig}_combined' url_only='yes'}
                    {/exp:jcogs_img:image:add_image}
                {/exp:jcogs_img:image:add_shadow}
            {/exp:jcogs_img:image:frame}
        {/exp:jcogs_img:image:shadow}
    {/exp:jcogs_img:image:black}
{/exp:jcogs_img:image:book}</code></pre>

**Environment Details:**
 - Version: 7.3.14
 - PHP Version 8.1

**Comment by bryannielsen** [comment]
@jcogs-design can you give me a little bit more code to show how these manipulations work?  Everything is run synchronously so I don't think the issue would be with how rapidly the calls are made but I'm wondering if you're using the filesystem's `write()` or `writeStream()` methods to save these manipulations or if you're using native PHP functions like `fwrite()` or `file_put_contents()`.  The filesystem library does make use of a cached adapter so if you call `ee('Filesystem')->exists($path)` and `$path` does not exist it caches that `false` response.  The cache for `$path` would be invalidated if you call `ee('Filesystem')->write($path)` though.  

All that to say if you can give me a simplified version of what that template code is doing and what filesystem methods (EE or native) it is calling that would help me better understand the situation.

**Comment by jcogs-design** [comment]
Hi Bryan.

Image writes its image files to disk using the appropriate php primitives - which depending theoretically on which graphics library is in use, but for current purposes assume GD2 methods. So things like `imagejpg` or `imagewebp`. 

I don't see that there is an elegant way I could replace these writes with calls to `ee('Filesystem')->write($path)` directly (as the calls occur deep within the Imagine image processing library) - but possibly I could put in a bogus call to the `->write($path)` prior to triggering the save operation.

Not sure how that would help though - since presumably the issue is arising after the image write not before.

As for a simplified version, I'll see what I can come up with, but doubt it (and anyhow that's not the issue here). 

But to explain a bit more clearly what the actual template code is doing: 

Start at the top and work in:

- `exp:jcogs_img:image:book` scales an image (of a book) to 450px high and proportionate width;
- `exp:jcogs_img:image:black` scales a black image to be 950px wide and 630px high;
- `exp:jcogs_img:image:shadow` generates a white image that has the same dimensions as the rescaled book image from step 1
- `exp:jcogs_img:image:frame` makes a copy of the `black` background image that is 2x the original's size
- `exp:jcogs_img:image:add_shadow` places the 'shadow' image created in step 2 onto the 'frame' image made in step 4
- `exp:jcogs_img:image:add_image` overlays the 'book' image from step 1 onto the output of the add_shadow operation in step 5, but offset by a set distance from the centre (to give the shadow effect)
- `exp:jcogs_img:image:combined` overlays the `add_image` composite onto the `black` image generated in step 2

It is possible that step 2 could be eliminated from this sequence.

The purpose of the sequence is to generate automatically the composite image that appears on the RHS of this kind of image 
![image](https://github.com/ExpressionEngine/ExpressionEngine/assets/13821249/fa2f812c-3884-4c68-abd3-2f1fe6983d7d)

**Comment by bryannielsen** [comment]
Yea this makes sense then, if all the file creation and manipulation is done using native PHP functions the EE filesystem library shouldn't be used for the `exists()` check because it won't know when there are any changes.  So if you don't need to support multiple filesystem drivers (like AWS S3, Digital Ocean Spaces, etc...) it should be fine to use PHP functions like `file_exists()` or `is_readable()` instead of the EE filesystem.  

If you do want to support those adapters though you would need to take the output from your manipulation and pass it through to the Filesystem library `write()` or `writeStream()` functions to store it with those adapters which would also register the change in our caching layer.  When we do manipulations in the core we're often working with a temporary file and then write it to the filesystem driver when we're done with the processing.  

So bottom line I think this is expected behavior because filesystem manipulation is happening outside of the filesystem library.

**Comment by jcogs-design** [comment]
What I want to do is ensure the add-on will continue to function should a user choose to use a Filesystem option in the EE instance. Image itself has no interaction with the EE Files sub-system - it simply gets file paths delivered to it from the EE template parsing system, which (if the path is a local one) it will attempt to open locally via php file system calls. Currently Image only writes files out to folders that live within the site's web-root (so that they can be retrieved by web clients).

I'm unclear from your answer if this then means I should or should not be using EE Filesystem calls to validate the presence of the provided local file, and / or I suppose write the files to the web-root located image cache directory.
